{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing the transcript structures of syntelogs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_file_leaf = \"/scratch/nadjafn/Atlantic_ASE/nf-potato-ase/tama_test_leaf/leaf_chr01.bed\"\n",
    "bed_file_tuber = \"/scratch/nadjafn/Atlantic_ASE/nf-potato-ase/tama_test_tuber/tuber_chr01.bed\"\n",
    "cluster_info_leaf = \"/scratch/nadjafn/Atlantic_ASE/nf-potato-ase/tama_test_leaf/leaf_chr01_trans_report.txt\"\n",
    "cluster_info_tuber = \"/scratch/nadjafn/Atlantic_ASE/nf-potato-ase/tama_test_tuber/tuber_chr01_trans_report.txt\"\n",
    "allele_info = \"/DKED/scratch/nadjafn/Atlantic_ASE/nf-potato-ase/results/04_ALLELECOUNTS/Synt_counts_with_info.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the bed file\n",
    "bed_filtered_list = []\n",
    "bed_list = []\n",
    "for bed_file, cluster_info in zip([bed_file_leaf, bed_file_tuber], [cluster_info_leaf, cluster_info_tuber]):\n",
    "    bed = pd.read_csv(bed_file, sep=\"\\t\", header=None, names=[\"haplotype_id\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"thickstart\", \"thickend\", \"itemRgb\", \"blockCount\", \"blockSizes\", \"blockStarts\"])\n",
    "    # add a colum with transcript id which is a split of the name after ;\n",
    "    bed[\"transcript_id\"] = bed[\"name\"].str.split(\";\").str[1]\n",
    "    clusters = pd.read_csv(cluster_info, sep=\"\\t\", usecols=[0, 1])\n",
    "    # merge the bed file with the cluster info\n",
    "    bed = bed.merge(clusters, on=\"transcript_id\", how=\"left\")\n",
    "    bed_unfiltered = bed\n",
    "    # Define a helper function to check if any two values in a series are within 80%\n",
    "    def check_within_threshold(values, threshold=0.8):\n",
    "        sorted_values = sorted(values, reverse= True)  # Sort the values for comparison\n",
    "        # drop NaN values\n",
    "        sorted_values = [x for x in sorted_values if str(x) != 'nan']\n",
    "        if len(sorted_values) > 1:\n",
    "            if  sorted_values[1]/sorted_values[0] > threshold:\n",
    "                return True\n",
    "        return False\n",
    "\n",
    "    # Process each haplotype group\n",
    "    result = []\n",
    "    for haplotype_id, group in bed.groupby(\"haplotype_id\"):\n",
    "        if check_within_threshold(group['num_clusters'].values, threshold=0.6):\n",
    "            # If two counts are within 80%, keep the row with the maximum 'blockCount'\n",
    "            row = group.loc[group['blockCount'].idxmax()]\n",
    "        else:\n",
    "            # Otherwise, keep the row with the maximum 'num_clusters'\n",
    "            row = group.loc[group['num_clusters'].idxmax()]\n",
    "        result.append(row)\n",
    "\n",
    "    # Combine the filtered rows into a new DataFrame\n",
    "    bed_filtered = pd.DataFrame(result)\n",
    "    bed_filtered = bed_filtered[bed_filtered[\"num_clusters\"] > 50]\n",
    "    bed_filtered_list.append(bed_filtered)\n",
    "    bed_list.append(bed)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickstart</th>\n",
       "      <th>thickend</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16297</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>0</td>\n",
       "      <td>4700</td>\n",
       "      <td>G3897;G3897.1</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>4700</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>361,97,92,132,186,164,79,129,128,327</td>\n",
       "      <td>0,575,754,929,1144,1488,1759,3754,4160,4373</td>\n",
       "      <td>G3897.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16298</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>0</td>\n",
       "      <td>5362</td>\n",
       "      <td>G3897;G3897.2</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>5362</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>12</td>\n",
       "      <td>34,240,97,92,132,186,164,79,129,128,408,453</td>\n",
       "      <td>0,121,575,754,929,1144,1488,1759,3754,4160,437...</td>\n",
       "      <td>G3897.2</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16299</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3</td>\n",
       "      <td>3819</td>\n",
       "      <td>G3897;G3897.3</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>3819</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>8</td>\n",
       "      <td>358,97,92,132,186,164,79,65</td>\n",
       "      <td>0,572,751,926,1141,1485,1756,3751</td>\n",
       "      <td>G3897.3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16300</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3</td>\n",
       "      <td>4284</td>\n",
       "      <td>G3897;G3897.4</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>4284</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>31,240,97,92,132,186,164,79,129,124</td>\n",
       "      <td>0,118,572,751,926,1141,1485,1756,3751,4157</td>\n",
       "      <td>G3897.4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16301</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3</td>\n",
       "      <td>4720</td>\n",
       "      <td>G3897;G3897.5</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>4720</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>358,97,92,132,186,164,79,129,128,347</td>\n",
       "      <td>0,572,751,926,1141,1485,1756,3751,4157,4370</td>\n",
       "      <td>G3897.5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16302</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>7</td>\n",
       "      <td>4700</td>\n",
       "      <td>G3897;G3897.6</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>4700</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>11</td>\n",
       "      <td>27,240,97,92,132,186,164,79,129,128,327</td>\n",
       "      <td>0,114,568,747,922,1137,1481,1752,3747,4153,4366</td>\n",
       "      <td>G3897.6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16303</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>7</td>\n",
       "      <td>5191</td>\n",
       "      <td>G3897;G3897.7</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>5191</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>12</td>\n",
       "      <td>27,240,97,92,132,186,164,79,129,128,408,282</td>\n",
       "      <td>0,114,568,747,922,1137,1481,1752,3747,4153,436...</td>\n",
       "      <td>G3897.7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16304</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>7</td>\n",
       "      <td>5346</td>\n",
       "      <td>G3897;G3897.8</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>5346</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>9</td>\n",
       "      <td>354,97,92,132,186,164,79,105,309</td>\n",
       "      <td>0,568,747,922,1137,1481,1752,3747,5030</td>\n",
       "      <td>G3897.8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16305</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>8</td>\n",
       "      <td>1915</td>\n",
       "      <td>G3897;G3897.9</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>8</td>\n",
       "      <td>1915</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>8</td>\n",
       "      <td>26,240,97,92,132,186,164,156</td>\n",
       "      <td>0,113,567,746,921,1136,1480,1751</td>\n",
       "      <td>G3897.9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16306</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>8</td>\n",
       "      <td>4948</td>\n",
       "      <td>G3897;G3897.10</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>8</td>\n",
       "      <td>4948</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>12</td>\n",
       "      <td>26,240,97,92,132,186,164,79,129,128,408,39</td>\n",
       "      <td>0,113,567,746,921,1136,1480,1751,3746,4152,436...</td>\n",
       "      <td>G3897.10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16307</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>12</td>\n",
       "      <td>5362</td>\n",
       "      <td>G3897;G3897.11</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>5362</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>11</td>\n",
       "      <td>349,97,92,132,186,164,79,129,128,408,453</td>\n",
       "      <td>0,563,742,917,1132,1476,1747,3742,4148,4361,4897</td>\n",
       "      <td>G3897.11</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16308</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>30</td>\n",
       "      <td>5385</td>\n",
       "      <td>G3897;G3897.12</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>30</td>\n",
       "      <td>5385</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>331,97,92,132,186,164,79,129,128,1012</td>\n",
       "      <td>0,545,724,899,1114,1458,1729,3724,4130,4343</td>\n",
       "      <td>G3897.12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16309</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>57</td>\n",
       "      <td>5274</td>\n",
       "      <td>G3897;G3897.13</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>57</td>\n",
       "      <td>5274</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>11</td>\n",
       "      <td>304,97,92,132,186,164,79,129,128,408,365</td>\n",
       "      <td>0,518,697,872,1087,1431,1702,3697,4103,4316,4852</td>\n",
       "      <td>G3897.13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16310</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3612</td>\n",
       "      <td>5362</td>\n",
       "      <td>G3897;G3897.14</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3612</td>\n",
       "      <td>5362</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>4</td>\n",
       "      <td>271,128,408,453</td>\n",
       "      <td>0,548,761,1297</td>\n",
       "      <td>G3897.14</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16311</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3858</td>\n",
       "      <td>4828</td>\n",
       "      <td>G3897;G3897.15</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3858</td>\n",
       "      <td>4828</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>3</td>\n",
       "      <td>25,128,455</td>\n",
       "      <td>0,302,515</td>\n",
       "      <td>G3897.15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16312</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>4535</td>\n",
       "      <td>5450</td>\n",
       "      <td>G3897;G3897.16</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>4535</td>\n",
       "      <td>5450</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>246,541</td>\n",
       "      <td>0,374</td>\n",
       "      <td>G3897.16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            haplotype_id  start   end  \\\n",
       "16297  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      0  4700   \n",
       "16298  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      0  5362   \n",
       "16299  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      3  3819   \n",
       "16300  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      3  4284   \n",
       "16301  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      3  4720   \n",
       "16302  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      7  4700   \n",
       "16303  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      7  5191   \n",
       "16304  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      7  5346   \n",
       "16305  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      8  1915   \n",
       "16306  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      8  4948   \n",
       "16307  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     12  5362   \n",
       "16308  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     30  5385   \n",
       "16309  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     57  5274   \n",
       "16310  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....   3612  5362   \n",
       "16311  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....   3858  4828   \n",
       "16312  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....   4535  5450   \n",
       "\n",
       "                 name  score strand  thickstart  thickend  itemRgb  \\\n",
       "16297   G3897;G3897.1     40      +           0      4700  255,0,0   \n",
       "16298   G3897;G3897.2     40      +           0      5362  255,0,0   \n",
       "16299   G3897;G3897.3     40      +           3      3819  255,0,0   \n",
       "16300   G3897;G3897.4     40      +           3      4284  255,0,0   \n",
       "16301   G3897;G3897.5     40      +           3      4720  255,0,0   \n",
       "16302   G3897;G3897.6     40      +           7      4700  255,0,0   \n",
       "16303   G3897;G3897.7     40      +           7      5191  255,0,0   \n",
       "16304   G3897;G3897.8     40      +           7      5346  255,0,0   \n",
       "16305   G3897;G3897.9     40      +           8      1915  255,0,0   \n",
       "16306  G3897;G3897.10     40      +           8      4948  255,0,0   \n",
       "16307  G3897;G3897.11     40      +          12      5362  255,0,0   \n",
       "16308  G3897;G3897.12     40      +          30      5385  255,0,0   \n",
       "16309  G3897;G3897.13     40      +          57      5274  255,0,0   \n",
       "16310  G3897;G3897.14     40      +        3612      5362  255,0,0   \n",
       "16311  G3897;G3897.15     40      +        3858      4828  255,0,0   \n",
       "16312  G3897;G3897.16     40      +        4535      5450  255,0,0   \n",
       "\n",
       "       blockCount                                   blockSizes  \\\n",
       "16297          10         361,97,92,132,186,164,79,129,128,327   \n",
       "16298          12  34,240,97,92,132,186,164,79,129,128,408,453   \n",
       "16299           8                  358,97,92,132,186,164,79,65   \n",
       "16300          10          31,240,97,92,132,186,164,79,129,124   \n",
       "16301          10         358,97,92,132,186,164,79,129,128,347   \n",
       "16302          11      27,240,97,92,132,186,164,79,129,128,327   \n",
       "16303          12  27,240,97,92,132,186,164,79,129,128,408,282   \n",
       "16304           9             354,97,92,132,186,164,79,105,309   \n",
       "16305           8                 26,240,97,92,132,186,164,156   \n",
       "16306          12   26,240,97,92,132,186,164,79,129,128,408,39   \n",
       "16307          11     349,97,92,132,186,164,79,129,128,408,453   \n",
       "16308          10        331,97,92,132,186,164,79,129,128,1012   \n",
       "16309          11     304,97,92,132,186,164,79,129,128,408,365   \n",
       "16310           4                              271,128,408,453   \n",
       "16311           3                                   25,128,455   \n",
       "16312           2                                      246,541   \n",
       "\n",
       "                                             blockStarts transcript_id  \\\n",
       "16297        0,575,754,929,1144,1488,1759,3754,4160,4373       G3897.1   \n",
       "16298  0,121,575,754,929,1144,1488,1759,3754,4160,437...       G3897.2   \n",
       "16299                  0,572,751,926,1141,1485,1756,3751       G3897.3   \n",
       "16300         0,118,572,751,926,1141,1485,1756,3751,4157       G3897.4   \n",
       "16301        0,572,751,926,1141,1485,1756,3751,4157,4370       G3897.5   \n",
       "16302    0,114,568,747,922,1137,1481,1752,3747,4153,4366       G3897.6   \n",
       "16303  0,114,568,747,922,1137,1481,1752,3747,4153,436...       G3897.7   \n",
       "16304             0,568,747,922,1137,1481,1752,3747,5030       G3897.8   \n",
       "16305                   0,113,567,746,921,1136,1480,1751       G3897.9   \n",
       "16306  0,113,567,746,921,1136,1480,1751,3746,4152,436...      G3897.10   \n",
       "16307   0,563,742,917,1132,1476,1747,3742,4148,4361,4897      G3897.11   \n",
       "16308        0,545,724,899,1114,1458,1729,3724,4130,4343      G3897.12   \n",
       "16309   0,518,697,872,1087,1431,1702,3697,4103,4316,4852      G3897.13   \n",
       "16310                                     0,548,761,1297      G3897.14   \n",
       "16311                                          0,302,515      G3897.15   \n",
       "16312                                              0,374      G3897.16   \n",
       "\n",
       "       num_clusters  \n",
       "16297             3  \n",
       "16298            87  \n",
       "16299             1  \n",
       "16300             1  \n",
       "16301             2  \n",
       "16302             2  \n",
       "16303             2  \n",
       "16304             1  \n",
       "16305             1  \n",
       "16306             1  \n",
       "16307            85  \n",
       "16308             3  \n",
       "16309             3  \n",
       "16310            34  \n",
       "16311             1  \n",
       "16312             1  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "bed_list[0][bed_list[0][\"haplotype_id\"].str.contains(\"Synt_1354_chr01_3G_x4\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickstart</th>\n",
       "      <th>thickend</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5773</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3</td>\n",
       "      <td>5222</td>\n",
       "      <td>G2257;G2257.1</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>5222</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>12</td>\n",
       "      <td>31,240,97,92,132,186,164,79,129,128,408,313</td>\n",
       "      <td>0,118,572,751,926,1141,1485,1756,3751,4157,437...</td>\n",
       "      <td>G2257.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5774</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3</td>\n",
       "      <td>5362</td>\n",
       "      <td>G2257;G2257.2</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>5362</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>12</td>\n",
       "      <td>31,240,97,92,132,186,164,79,129,128,408,453</td>\n",
       "      <td>0,118,572,751,926,1141,1485,1756,3751,4157,437...</td>\n",
       "      <td>G2257.2</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5775</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>3</td>\n",
       "      <td>5362</td>\n",
       "      <td>G2257;G2257.3</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>5362</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>11</td>\n",
       "      <td>358,97,92,132,186,164,79,129,128,408,453</td>\n",
       "      <td>0,572,751,926,1141,1485,1756,3751,4157,4370,4906</td>\n",
       "      <td>G2257.3</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>55</td>\n",
       "      <td>5362</td>\n",
       "      <td>G2257;G2257.4</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>55</td>\n",
       "      <td>5362</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>306,97,92,132,186,164,79,129,128,989</td>\n",
       "      <td>0,520,699,874,1089,1433,1704,3699,4105,4318</td>\n",
       "      <td>G2257.4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5777</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>56</td>\n",
       "      <td>4522</td>\n",
       "      <td>G2257;G2257.5</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>56</td>\n",
       "      <td>4522</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>305,97,92,132,186,164,79,129,128,149</td>\n",
       "      <td>0,519,698,873,1088,1432,1703,3698,4104,4317</td>\n",
       "      <td>G2257.5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5778</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>60</td>\n",
       "      <td>4677</td>\n",
       "      <td>G2257;G2257.6</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>60</td>\n",
       "      <td>4677</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>10</td>\n",
       "      <td>301,97,92,132,186,164,79,129,128,304</td>\n",
       "      <td>0,515,694,869,1084,1428,1699,3694,4100,4313</td>\n",
       "      <td>G2257.6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5779</th>\n",
       "      <td>Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....</td>\n",
       "      <td>65</td>\n",
       "      <td>5113</td>\n",
       "      <td>G2257;G2257.7</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>65</td>\n",
       "      <td>5113</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>11</td>\n",
       "      <td>296,97,92,132,186,164,79,129,128,408,204</td>\n",
       "      <td>0,510,689,864,1079,1423,1694,3689,4095,4308,4844</td>\n",
       "      <td>G2257.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           haplotype_id  start   end  \\\n",
       "5773  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      3  5222   \n",
       "5774  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      3  5362   \n",
       "5775  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....      3  5362   \n",
       "5776  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     55  5362   \n",
       "5777  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     56  4522   \n",
       "5778  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     60  4677   \n",
       "5779  Synt_1354_chr01_3G_x4|hap3_Soltu.DM.01G020170....     65  5113   \n",
       "\n",
       "               name  score strand  thickstart  thickend  itemRgb  blockCount  \\\n",
       "5773  G2257;G2257.1     40      +           3      5222  255,0,0          12   \n",
       "5774  G2257;G2257.2     40      +           3      5362  255,0,0          12   \n",
       "5775  G2257;G2257.3     40      +           3      5362  255,0,0          11   \n",
       "5776  G2257;G2257.4     40      +          55      5362  255,0,0          10   \n",
       "5777  G2257;G2257.5     40      +          56      4522  255,0,0          10   \n",
       "5778  G2257;G2257.6     40      +          60      4677  255,0,0          10   \n",
       "5779  G2257;G2257.7     40      +          65      5113  255,0,0          11   \n",
       "\n",
       "                                       blockSizes  \\\n",
       "5773  31,240,97,92,132,186,164,79,129,128,408,313   \n",
       "5774  31,240,97,92,132,186,164,79,129,128,408,453   \n",
       "5775     358,97,92,132,186,164,79,129,128,408,453   \n",
       "5776         306,97,92,132,186,164,79,129,128,989   \n",
       "5777         305,97,92,132,186,164,79,129,128,149   \n",
       "5778         301,97,92,132,186,164,79,129,128,304   \n",
       "5779     296,97,92,132,186,164,79,129,128,408,204   \n",
       "\n",
       "                                            blockStarts transcript_id  \\\n",
       "5773  0,118,572,751,926,1141,1485,1756,3751,4157,437...       G2257.1   \n",
       "5774  0,118,572,751,926,1141,1485,1756,3751,4157,437...       G2257.2   \n",
       "5775   0,572,751,926,1141,1485,1756,3751,4157,4370,4906       G2257.3   \n",
       "5776        0,520,699,874,1089,1433,1704,3699,4105,4318       G2257.4   \n",
       "5777        0,519,698,873,1088,1432,1703,3698,4104,4317       G2257.5   \n",
       "5778        0,515,694,869,1084,1428,1699,3694,4100,4313       G2257.6   \n",
       "5779   0,510,689,864,1079,1423,1694,3689,4095,4308,4844       G2257.7   \n",
       "\n",
       "      num_clusters  \n",
       "5773             3  \n",
       "5774            29  \n",
       "5775            42  \n",
       "5776             2  \n",
       "5777             1  \n",
       "5778             2  \n",
       "5779             1  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_list[1][bed_list[1][\"haplotype_id\"].str.contains(\"Synt_1354_chr01_3G_x4\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         haplotype_id  blockCount_leaf  \\\n",
      "30  Synt_3005_chr01_2G_x4|hap2_Soltu.DM.01G042380....                9   \n",
      "45  Synt_3364_chr01_1G_x4|hap1_Soltu.DM.01G047280....               10   \n",
      "48  Synt_3364_chr01_4G_x4|hap4_Soltu.DM.01G047280....               11   \n",
      "\n",
      "   transcript_id_leaf  num_clusters_leaf  blockCount_tuber  \\\n",
      "30           G3223.12                 61                 8   \n",
      "45            G1958.5                167                 9   \n",
      "48            G6467.7                158                 9   \n",
      "\n",
      "   transcript_id_tuber  num_clusters_tuber  \n",
      "30             G1853.1                 223  \n",
      "45             G1133.1                 128  \n",
      "48             G3818.1                 178  \n"
     ]
    }
   ],
   "source": [
    "# we merge the two dataframes\n",
    "bed = bed_filtered_list[0].merge(bed_filtered_list[1], on=\"haplotype_id\", how=\"inner\", suffixes=(\"_leaf\", \"_tuber\"))\n",
    "# drop the columns that are not needed\n",
    "bed.drop(columns=[\"start_leaf\", \"end_leaf\", \"start_tuber\", \"end_tuber\", \"name_leaf\", \"name_tuber\", \"score_leaf\", \"score_tuber\", \"strand_leaf\", \"strand_tuber\", \"thickstart_leaf\", \"thickstart_tuber\", \"thickend_leaf\", \"thickend_tuber\", \"itemRgb_leaf\", \"itemRgb_tuber\", \"blockSizes_leaf\", \"blockSizes_tuber\", \"blockStarts_leaf\", \"blockStarts_tuber\"], inplace=True)\n",
    "# print the rows that have a difference in the number of blockCounts\n",
    "print(bed[bed[\"blockCount_leaf\"] != bed[\"blockCount_tuber\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype_id</th>\n",
       "      <th>blockCount_leaf</th>\n",
       "      <th>transcript_id_leaf</th>\n",
       "      <th>num_clusters_leaf</th>\n",
       "      <th>blockCount_tuber</th>\n",
       "      <th>transcript_id_tuber</th>\n",
       "      <th>num_clusters_tuber</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Synt_125_chr01_1G_x4|hap1_Soltu.DM.01G001830.1...</td>\n",
       "      <td>5</td>\n",
       "      <td>G843.2</td>\n",
       "      <td>87</td>\n",
       "      <td>5</td>\n",
       "      <td>G503.1</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Synt_125_chr01_2G_x4|hap2_Soltu.DM.01G001830.1...</td>\n",
       "      <td>5</td>\n",
       "      <td>G2135.1</td>\n",
       "      <td>82</td>\n",
       "      <td>5</td>\n",
       "      <td>G1228.1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Synt_125_chr01_3G_x4|hap3_Soltu.DM.01G001830.1...</td>\n",
       "      <td>5</td>\n",
       "      <td>G3618.1</td>\n",
       "      <td>62</td>\n",
       "      <td>5</td>\n",
       "      <td>G2099.1</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Synt_125_chr01_4G_x4|hap4_Soltu.DM.01G001830.1...</td>\n",
       "      <td>5</td>\n",
       "      <td>G5199.1</td>\n",
       "      <td>93</td>\n",
       "      <td>5</td>\n",
       "      <td>G3054.1</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Synt_1530_chr01_2G_x3|hap2_Soltu.DM.01G022570....</td>\n",
       "      <td>3</td>\n",
       "      <td>G2537.3</td>\n",
       "      <td>184</td>\n",
       "      <td>3</td>\n",
       "      <td>G1420.1</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>no_syntelog_chr01_3G_x|hap3_Soltu.DM.01G044610...</td>\n",
       "      <td>5</td>\n",
       "      <td>G4729.1</td>\n",
       "      <td>121</td>\n",
       "      <td>5</td>\n",
       "      <td>G2782.1</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>no_syntelog_chr01_3G_x|hap3_Soltu.DM.10G013440...</td>\n",
       "      <td>2</td>\n",
       "      <td>G3864.3</td>\n",
       "      <td>219</td>\n",
       "      <td>2</td>\n",
       "      <td>G2236.1</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>no_syntelog_chr01_4G_x|hap4_Soltu.DM.01G027390...</td>\n",
       "      <td>13</td>\n",
       "      <td>G5749.1</td>\n",
       "      <td>99</td>\n",
       "      <td>13</td>\n",
       "      <td>G3360.2</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>no_syntelog_chr01_4G_x|hap4_Soltu.DM.01G044610...</td>\n",
       "      <td>5</td>\n",
       "      <td>G6342.1</td>\n",
       "      <td>186</td>\n",
       "      <td>5</td>\n",
       "      <td>G3733.1</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>no_syntelog_chr01_4G_x|hap4_Soltu.DM.03G033940...</td>\n",
       "      <td>7</td>\n",
       "      <td>G5153.1</td>\n",
       "      <td>78</td>\n",
       "      <td>7</td>\n",
       "      <td>G3030.3</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          haplotype_id  blockCount_leaf  \\\n",
       "0    Synt_125_chr01_1G_x4|hap1_Soltu.DM.01G001830.1...                5   \n",
       "1    Synt_125_chr01_2G_x4|hap2_Soltu.DM.01G001830.1...                5   \n",
       "2    Synt_125_chr01_3G_x4|hap3_Soltu.DM.01G001830.1...                5   \n",
       "3    Synt_125_chr01_4G_x4|hap4_Soltu.DM.01G001830.1...                5   \n",
       "4    Synt_1530_chr01_2G_x3|hap2_Soltu.DM.01G022570....                3   \n",
       "..                                                 ...              ...   \n",
       "168  no_syntelog_chr01_3G_x|hap3_Soltu.DM.01G044610...                5   \n",
       "169  no_syntelog_chr01_3G_x|hap3_Soltu.DM.10G013440...                2   \n",
       "170  no_syntelog_chr01_4G_x|hap4_Soltu.DM.01G027390...               13   \n",
       "171  no_syntelog_chr01_4G_x|hap4_Soltu.DM.01G044610...                5   \n",
       "172  no_syntelog_chr01_4G_x|hap4_Soltu.DM.03G033940...                7   \n",
       "\n",
       "    transcript_id_leaf  num_clusters_leaf  blockCount_tuber  \\\n",
       "0               G843.2                 87                 5   \n",
       "1              G2135.1                 82                 5   \n",
       "2              G3618.1                 62                 5   \n",
       "3              G5199.1                 93                 5   \n",
       "4              G2537.3                184                 3   \n",
       "..                 ...                ...               ...   \n",
       "168            G4729.1                121                 5   \n",
       "169            G3864.3                219                 2   \n",
       "170            G5749.1                 99                13   \n",
       "171            G6342.1                186                 5   \n",
       "172            G5153.1                 78                 7   \n",
       "\n",
       "    transcript_id_tuber  num_clusters_tuber  \n",
       "0                G503.1                  50  \n",
       "1               G1228.1                  33  \n",
       "2               G2099.1                  45  \n",
       "3               G3054.1                  50  \n",
       "4               G1420.1                 128  \n",
       "..                  ...                 ...  \n",
       "168             G2782.1                  62  \n",
       "169             G2236.1                  75  \n",
       "170             G3360.2                  31  \n",
       "171             G3733.1                  61  \n",
       "172             G3030.3                  43  \n",
       "\n",
       "[173 rows x 7 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        Synt_id haplotype  \\\n",
      "0      Soltu.Atl_v3.01_1G016780         1   \n",
      "1      Soltu.Atl_v3.01_1G018980         1   \n",
      "2            Soltu.DM.10G009940         1   \n",
      "3      Soltu.Atl_v3.01_1G022210         1   \n",
      "4      Soltu.Atl_v3.01_1G023000         1   \n",
      "...                         ...       ...   \n",
      "44019        Soltu.DM.12G029210         4   \n",
      "44020        Soltu.DM.12G029220         4   \n",
      "44021        Soltu.DM.12G029230         4   \n",
      "44022        Soltu.DM.12G029240         4   \n",
      "44023        Soltu.DM.12G029300         4   \n",
      "\n",
      "                                            haplotype_id  \n",
      "0      Synt_1744_chr01_1G_x4|Soltu.Atl_v3.01_1G016780...  \n",
      "1      Synt_1970_chr01_1G_x4|Soltu.Atl_v3.01_1G018980...  \n",
      "2      Synt_20601_chr08_1G_x8|hap1_Soltu.DM.10G009940...  \n",
      "3      Synt_2300_chr01_1G_x4|Soltu.Atl_v3.01_1G022210...  \n",
      "4      Synt_2499_chr01_1G_x4|Soltu.Atl_v3.01_1G023000...  \n",
      "...                                                  ...  \n",
      "44019  no_syntelog_chr12_4G_x|hap4_Soltu.DM.12G029210...  \n",
      "44020  no_syntelog_chr12_4G_x|hap4_Soltu.DM.12G029220...  \n",
      "44021  Synt_30370_chr12_4G_x4|hap4_Soltu.DM.12G029230...  \n",
      "44022  Synt_30371_chr12_4G_x3|hap4_Soltu.DM.12G029240...  \n",
      "44023  no_syntelog_chr12_4G_x|hap4_Soltu.DM.12G029300...  \n",
      "\n",
      "[44024 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# pivot the allele table\n",
    "# 1. read the allele table\n",
    "df = pd.read_csv('/scratch/nadjafn/Atlantic_ASE/output/AlleleFinder/allelefinder/allele.adjusted.txt', sep='\\t')\n",
    "# select the rows that have exactly one entry in 'A', 'B', 'C', 'D' columns. so no , and no missing values\n",
    "df = df[(df['Allele A'].str.count(',') == 0) & (df['Allele B'].str.count(',') == 0) & (df['Allele C'].str.count(',') == 0) & (df['Allele D'].str.count(',') == 0)]\n",
    "# reset the index\n",
    "df = df.reset_index(drop=True)\n",
    "# Add a id column to the dataframe\n",
    "df['Synt_id'] = df.index\n",
    "# extract the chromosome from all the allele columns\n",
    "df['Chr_a'] = df['Allele A'].str.extract(r'(chr\\d+)')\n",
    "df['Chr_b'] = df['Allele B'].str.extract(r'(chr\\d+)')\n",
    "df['Chr_c'] = df['Allele C'].str.extract(r'(chr\\d+)')\n",
    "df['Chr_d'] = df['Allele D'].str.extract(r'(chr\\d+)')\n",
    "# filter that have the same chromomsome in all the alleles\n",
    "df = df[(df['Chr_a'] == df['Chr_b']) & (df['Chr_a'] == df['Chr_c']) & (df['Chr_a'] == df['Chr_d'])]\n",
    "df['Synt_id'] = df['Allele A'].str.extract(r'(Soltu.DM.\\d+G\\d+)')\n",
    "\n",
    "# If the first pattern is not found, extract the second pattern and fill in the missing values\n",
    "df['Synt_id'] = df['Synt_id'].fillna(df['Allele A'].str.extract(r'(Soltu.Atl_v3.01_\\d+G\\d*)')[0])\n",
    "\n",
    "# 2. pivot the table so that each row is a single allele\n",
    "# 2.1. select the columns that we want to pivot\n",
    "df_synt = df[['Synt_id', 'Allele A', 'Allele B', 'Allele C', 'Allele D']]\n",
    "# 2.2. rename the columns\n",
    "df_synt.columns = ['Synt_id', '1', '2', '3', '4']\n",
    "# 2.3. melt the table\n",
    "df_synt = pd.melt(df_synt, id_vars=['Synt_id'], value_vars=['1', '2', '3', '4'], var_name='haplotype', value_name='haplotype_id')\n",
    "# replace Synt_id with a part from haplotype id\n",
    "# Extract the first pattern\n",
    "print(df_synt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Column not found: blockCount'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m merged \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(df_synt, bed, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhaplotype_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minner\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# group by synt_id and count the number of unuqie blockCount values\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m count \u001b[38;5;241m=\u001b[39m \u001b[43mmerged\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mSynt_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mblockCount\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mnunique()\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# make a barplot of the count\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# count the frequency of the counts\u001b[39;00m\n\u001b[1;32m      6\u001b[0m plt \u001b[38;5;241m=\u001b[39m count\u001b[38;5;241m.\u001b[39mvalue_counts()\u001b[38;5;241m.\u001b[39mplot(kind\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbar\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m/DKED/scratch/nadjafn/Phasing/ASE-tools-benchmark/conda/expressionMatrix-613c97c23a72e82e6de2bbc3b086d489/lib/python3.11/site-packages/pandas/core/groupby/generic.py:1961\u001b[0m, in \u001b[0;36mDataFrameGroupBy.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1954\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mtuple\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(key) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   1955\u001b[0m     \u001b[38;5;66;03m# if len == 1, then it becomes a SeriesGroupBy and this is actually\u001b[39;00m\n\u001b[1;32m   1956\u001b[0m     \u001b[38;5;66;03m# valid syntax, so don't raise\u001b[39;00m\n\u001b[1;32m   1957\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1958\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot subset columns with a tuple with more than one element. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1959\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUse a list instead.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1960\u001b[0m     )\n\u001b[0;32m-> 1961\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__getitem__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/DKED/scratch/nadjafn/Phasing/ASE-tools-benchmark/conda/expressionMatrix-613c97c23a72e82e6de2bbc3b086d489/lib/python3.11/site-packages/pandas/core/base.py:244\u001b[0m, in \u001b[0;36mSelectionMixin.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    243\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj:\n\u001b[0;32m--> 244\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mColumn not found: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    245\u001b[0m     ndim \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj[key]\u001b[38;5;241m.\u001b[39mndim\n\u001b[1;32m    246\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_gotitem(key, ndim\u001b[38;5;241m=\u001b[39mndim)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Column not found: blockCount'"
     ]
    }
   ],
   "source": [
    "merged = pd.merge(df_synt, bed, on=\"haplotype_id\", how=\"inner\")\n",
    "# group by synt_id and count the number of unuqie blockCount values\n",
    "count = merged.groupby(\"Synt_id\")[\"blockCount\"].nunique()\n",
    "# make a barplot of the count\n",
    "# count the frequency of the counts\n",
    "plt = count.value_counts().plot(kind='bar')\n",
    "plt.set_xlabel(\"Number of exon numbers per syntelog\")  # Correct syntax for x-axis label\n",
    "plt.set_ylabel(\"Frequency\")  # Optionally, add a y-axis label\n",
    "plt.set_title(\"Distribution of Exon Numbers per Syntelog\")  # Optionally, add a title\n",
    " # Display the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Synt_id\n",
      "Soltu.DM.01G000050    2\n",
      "Soltu.DM.01G000120    2\n",
      "Soltu.DM.01G002580    2\n",
      "Soltu.DM.01G002690    2\n",
      "Soltu.DM.01G005580    2\n",
      "Soltu.DM.01G019020    2\n",
      "Soltu.DM.01G019780    2\n",
      "Soltu.DM.01G028100    2\n",
      "Soltu.DM.01G031330    2\n",
      "Soltu.DM.01G039510    2\n",
      "Soltu.DM.01G045870    2\n",
      "Soltu.DM.01G051500    2\n",
      "Name: blockCount, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# print where count > 1\n",
    "print(count[count > 1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Synt_id</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>haplotype_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickstart</th>\n",
       "      <th>thickend</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Soltu.DM.01G051500</td>\n",
       "      <td>2</td>\n",
       "      <td>Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....</td>\n",
       "      <td>221</td>\n",
       "      <td>2865</td>\n",
       "      <td>G3033;G3033.2</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>221</td>\n",
       "      <td>2865</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>3</td>\n",
       "      <td>651,188,494</td>\n",
       "      <td>0,1024,2150</td>\n",
       "      <td>G3033.2</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>Soltu.DM.01G051500</td>\n",
       "      <td>4</td>\n",
       "      <td>Synt_3676_chr01_4G_x4|hap4_Soltu.DM.01G051500....</td>\n",
       "      <td>169</td>\n",
       "      <td>2891</td>\n",
       "      <td>G5691;G5691.1</td>\n",
       "      <td>40</td>\n",
       "      <td>+</td>\n",
       "      <td>169</td>\n",
       "      <td>2891</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>1264,520</td>\n",
       "      <td>0,2202</td>\n",
       "      <td>G5691.1</td>\n",
       "      <td>94.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Synt_id haplotype  \\\n",
       "105  Soltu.DM.01G051500         2   \n",
       "217  Soltu.DM.01G051500         4   \n",
       "\n",
       "                                          haplotype_id  start   end  \\\n",
       "105  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    221  2865   \n",
       "217  Synt_3676_chr01_4G_x4|hap4_Soltu.DM.01G051500....    169  2891   \n",
       "\n",
       "              name  score strand  thickstart  thickend  itemRgb  blockCount  \\\n",
       "105  G3033;G3033.2     40      +         221      2865  255,0,0           3   \n",
       "217  G5691;G5691.1     40      +         169      2891  255,0,0           2   \n",
       "\n",
       "      blockSizes  blockStarts transcript_id  num_clusters  \n",
       "105  651,188,494  0,1024,2150       G3033.2          67.0  \n",
       "217     1264,520       0,2202       G5691.1          94.0  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged[merged[\"Synt_id\"] == \"Soltu.DM.01G051500\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            haplotype_id  start   end  \\\n",
      "10270  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    200  2777   \n",
      "10271  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    221  2865   \n",
      "10272  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    224  2633   \n",
      "10273  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    236  2902   \n",
      "10274  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    417  2582   \n",
      "10275  Synt_3676_chr01_2G_x4|hap2_Soltu.DM.01G051500....    877  2988   \n",
      "\n",
      "                name  score strand  thickstart  thickend  itemRgb  blockCount  \\\n",
      "10270  G3033;G3033.1     40      +         200      2777  255,0,0           2   \n",
      "10271  G3033;G3033.2     40      +         221      2865  255,0,0           3   \n",
      "10272  G3033;G3033.3     40      +         224      2633  255,0,0           2   \n",
      "10273  G3033;G3033.4     40      +         236      2902  255,0,0           3   \n",
      "10274  G3033;G3033.5     40      +         417      2582  255,0,0           2   \n",
      "10275  G3033;G3033.6     40      +         877      2988  255,0,0           2   \n",
      "\n",
      "        blockSizes  blockStarts transcript_id  num_clusters  \n",
      "10270     1233,406       0,2171       G3033.1           1.0  \n",
      "10271  651,188,494  0,1024,2150       G3033.2          67.0  \n",
      "10272     1209,262       0,2147       G3033.3           9.0  \n",
      "10273   64,903,531   0,294,2135       G3033.4           1.0  \n",
      "10274     1016,211       0,1954       G3033.5           1.0  \n",
      "10275      556,617       0,1494       G3033.6          17.0  \n"
     ]
    }
   ],
   "source": [
    "# print merged where Synt_id == Soltu.DM.01G044600\n",
    "print(bed_unfiltered[bed_unfiltered[\"haplotype_id\"].str.contains(\"hap2_Soltu.DM.01G051500\")])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
